File: python/test/mod_test_flyer_json.py
"""Contract tests for save_flyer_json: the call, the file it writes, and its documented errors.

The values themselves are tested in Fortran (test/mod_test_flyer_json.F90); here the file is
compared with the same hand-written expected file and read back by an independent JSON parser.
"""
import sys
import os
import json
import math
import tempfile

import numpy as np

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))
# the generated package loads the shared library itself

from tensor_omics import save_flyer_json
from tensor_omics.error_handling import ToxError, ERR_NAN_INF, ERR_INVALID_INPUT, ERR_FILE_OPEN
from test_helpers import run_all_tests, assert_error

EXPECTED_FILE = os.path.join(os.path.dirname(__file__), '..', '..', 'test', 'test_files', 'flyer_expected.json')


def reference_data():
    """The data set of the Fortran suite and of the expected file, one gene per row before transposing."""
    return dict(
        expression_vectors=np.array([[1.0, 0.5, -2.0],
                                     [0.0, -0.0, 0.25],
                                     [1024.0, -0.125, 3.0],
                                     [1.5, 2.5, -3.5],
                                     [0.75, 8.0, -16.0]]).T,
        family_centroids=np.array([[0.375, 4.25, -8.125],
                                   [512.5, 0.1875, 0.5],
                                   [-0.0, 0.0, 0.0625]]).T,
        gene_to_fam=[2, 1, 2, 0, 1],
        is_outlier=[False, True, False, False, True],
        axis_labels=["liver", "brain", "heart"],
        family_ids=["F1", "F2", "F3"],
        gene_ids=["g1", "g2", "g\u00e8ne", "g4", "g5"],
        gene_species=["human", "mouse", "human", "fly", 'say "hi"'],
        gene_types=["ortholog", "paralog", "ortholog", "ortholog", "a\\b"],
    )


def assert_tox_error(call, code, argument, msg):
    """Fail unless `call` raises a ToxError with this code, naming this argument."""
    try:
        call()
    except ToxError as error:
        assert error.code == code, f"{msg}: expected code {code}, got {error.code}: {error}"
        assert error.argument == argument, f"{msg}: expected argument '{argument}', got '{error.argument}': {error}"
        assert f"'{argument}'" in str(error), f"{msg}: the message does not name '{argument}': {error}"
    else:
        raise AssertionError(f"{msg}: nothing was raised")


def test_reference_file_matches_expected():
    with tempfile.TemporaryDirectory() as directory:
        path = os.path.join(directory, "flyer.json")
        save_flyer_json(filename=path, **reference_data())
        with open(path, "rb") as written, open(EXPECTED_FILE, "rb") as expected:
            assert written.read() == expected.read(), "the file differs from the expected file"


def test_file_parses_with_every_field():
    with tempfile.TemporaryDirectory() as directory:
        path = os.path.join(directory, "flyer.json")
        save_flyer_json(filename=path, **reference_data())
        with open(path, encoding="utf-8") as stream:
            document = json.load(stream)

    assert document["kind"] == "tox-flyer" and document["version"] == 1, "kind and version"
    assert document["tissues"] == ["liver", "brain", "heart"], "tissues"

    families = document["families"]
    assert [family["family"] for family in families] == ["F1", "F2", "F3"], "family order"
    assert [family["gene_indices"] for family in families] == [[2, 5], [1, 3], []], "1-based gene indices"
    assert families[0]["centroid"] == [0.375, 4.25, -8.125], "centroid values"
    assert math.copysign(1.0, families[2]["centroid"][0]) == -1.0, "a negative zero keeps its sign"

    genes = document["genes"]
    assert len(genes) == 5, "one object per gene"
    for gene in genes:
        assert set(gene) == {"coordinates", "id", "family", "species", "is_outlier", "type"}, "all six keys"
    assert genes[2]["coordinates"] == [1024.0, -0.125, 3.0], "coordinates"
    assert [gene["id"] for gene in genes] == ["g1", "g2", "g\u00e8ne", "g4", "g5"], "ids, non-ASCII included"
    assert [gene["family"] for gene in genes] == ["F2", "F1", "F2", None, "F1"], "family ids, null for none"
    assert [gene["is_outlier"] for gene in genes] == [False, True, False, False, True], "outlier flags"
    assert all(type(gene["is_outlier"]) is bool for gene in genes), "outlier flags are JSON booleans"
    assert genes[4]["species"] == 'say "hi"' and genes[4]["type"] == "a\\b", "escaped strings"


def test_non_ascii_ids_round_trip():
    data = reference_data()
    data["gene_ids"] = ["基因", "\U0001F600", "été", "∅", "g5"]
    data["family_ids"] = ["族1", "F2", "F3"]
    with tempfile.TemporaryDirectory() as directory:
        path = os.path.join(directory, "flyer.json")
        save_flyer_json(filename=path, **data)
        with open(path, encoding="utf-8") as stream:
            document = json.load(stream)
    assert [gene["id"] for gene in document["genes"]] == data["gene_ids"], "non-ASCII ids read back"
    assert document["genes"][1]["family"] == "族1", "non-ASCII family id read back"


def test_documented_errors_name_the_argument():
    with tempfile.TemporaryDirectory() as directory:
        path = os.path.join(directory, "flyer.json")

        data = reference_data()
        data["expression_vectors"][1, 2] = np.nan
        assert_tox_error(lambda: save_flyer_json(filename=path, **data), ERR_NAN_INF, "expression_vectors", "NaN")

        data = reference_data()
        data["family_centroids"][0, 1] = np.inf
        assert_tox_error(lambda: save_flyer_json(filename=path, **data), ERR_NAN_INF, "family_centroids", "Inf")

        data = reference_data()
        data["gene_to_fam"] = [2, 1, 4, 0, 1]
        assert_tox_error(lambda: save_flyer_json(filename=path, **data), ERR_INVALID_INPUT, "gene_to_fam",
                         "family index above the number of families")

        data = reference_data()
        data["axis_labels"] = ["liver", "brain", "liver"]
        assert_tox_error(lambda: save_flyer_json(filename=path, **data), ERR_INVALID_INPUT, "axis_labels",
                         "repeated axis label")

        data = reference_data()
        data["gene_ids"] = ["g1", "g2", "g1", "g4", "g5"]
        assert_tox_error(lambda: save_flyer_json(filename=path, **data), ERR_INVALID_INPUT, "gene_ids",
                         "repeated gene id")

        data = reference_data()
        data["family_ids"] = ["F1", "", "F3"]
        assert_tox_error(lambda: save_flyer_json(filename=path, **data), ERR_INVALID_INPUT, "family_ids",
                         "empty family id")

        assert not os.path.exists(path), "a refused call creates no file"


def test_existing_file_is_refused():
    with tempfile.TemporaryDirectory() as directory:
        path = os.path.join(directory, "flyer.json")
        with open(path, "w") as stream:
            stream.write("keep me")
        assert_tox_error(lambda: save_flyer_json(filename=path, **reference_data()), ERR_FILE_OPEN, "filename",
                         "existing file")
        with open(path) as stream:
            assert stream.read() == "keep me", "the existing file is untouched"


def test_shape_mismatch_is_refused_before_the_call():
    data = reference_data()
    data["gene_ids"] = ["g1", "g2", "g3", "g4"]
    with tempfile.TemporaryDirectory() as directory:
        path = os.path.join(directory, "flyer.json")
        assert_error(lambda: save_flyer_json(filename=path, **data), "gene_ids shorter than the genes")
        assert not os.path.exists(path), "a refused call creates no file"


if __name__ == "__main__":
    run_all_tests(list(globals().values()))

File: r/test/mod_test_flyer_json.R
source("r/load_tensor_omics.R")
source("r/test_helpers.R")

# Contract tests for save_flyer_json: the call, the file it writes, and its documented errors.
# R has no JSON parser without extra packages, so the file is compared with the hand-written
# expected file of the Fortran suite, which holds the same data set.

EXPECTED_FILE <- "test/test_files/flyer_expected.json"

reference_data <- function() {
  # R's byte compiler pools constants it finds identical(), and identical(0, -0) is TRUE, so a
  # literal -0.0 may come out as 0.0 or turn a 0.0 negative. The negative zeros are therefore
  # computed at run time.
  negative_zero <- -numeric(1)
  expression_vectors <- matrix(c(1.0, 0.5, -2.0,
                                 0.0, 0.0, 0.25,
                                 1024.0, -0.125, 3.0,
                                 1.5, 2.5, -3.5,
                                 0.75, 8.0, -16.0), nrow = 3L)
  expression_vectors[2, 2] <- negative_zero
  family_centroids <- matrix(c(0.375, 4.25, -8.125,
                               512.5, 0.1875, 0.5,
                               0.0, 0.0, 0.0625), nrow = 3L)
  family_centroids[1, 3] <- negative_zero
  list(
    expression_vectors = expression_vectors,
    family_centroids = family_centroids,
    gene_to_fam = c(2L, 1L, 2L, 0L, 1L),
    is_outlier = c(FALSE, TRUE, FALSE, FALSE, TRUE),
    axis_labels = c("liver", "brain", "heart"),
    family_ids = c("F1", "F2", "F3"),
    gene_ids = c("g1", "g2", "g\u00e8ne", "g4", "g5"),
    gene_species = c("human", "mouse", "human", "fly", "say \"hi\""),
    gene_types = c("ortholog", "paralog", "ortholog", "ortholog", "a\\b")
  )
}

write_flyer <- function(data, path) {
  do.call(save_flyer_json, c(data, list(filename = path)))
}

# Fail unless `expr` raises a tox_error with this code, naming this argument.
assert_tox_error <- function(expr, code, argument, msg) {
  err <- tryCatch({ expr; NULL }, error = function(e) e)
  if (is.null(err)) stop(msg, ": nothing was raised", call. = FALSE)
  if (!inherits(err, "tox_error") || !identical(as.integer(err$code), as.integer(code)))
    stop(msg, ": expected code ", code, ", got: ", conditionMessage(err), call. = FALSE)
  if (!identical(err$argument, argument))
    stop(msg, ": expected argument '", argument, "', got '", err$argument, "'", call. = FALSE)
  if (!grepl(sprintf("'%s'", argument), conditionMessage(err), fixed = TRUE))
    stop(msg, ": the message does not name '", argument, "'", call. = FALSE)
  invisible(TRUE)
}

test_reference_file_matches_expected <- function() {
  path <- tempfile(fileext = ".json")
  on.exit(if (file.exists(path)) file.remove(path))
  write_flyer(reference_data(), path)

  written <- readBin(path, "raw", n = file.size(path))
  expected <- readBin(EXPECTED_FILE, "raw", n = file.size(EXPECTED_FILE))
  assert_true(identical(written, expected), "the file differs from the expected file, byte for byte")
  lines <- readLines(path, encoding = "UTF-8", warn = FALSE)
  assert_true(identical(lines, readLines(EXPECTED_FILE, encoding = "UTF-8", warn = FALSE)),
              "the file differs from the expected file, line for line")
  assert_equal_int(length(lines), 1L, "the document is a single line")
}

test_documented_errors_name_the_argument <- function() {
  path <- tempfile(fileext = ".json")
  on.exit(if (file.exists(path)) file.remove(path))

  data <- reference_data(); data$expression_vectors[2, 3] <- NaN
  assert_tox_error(write_flyer(data, path), ERR_NAN_INF, "expression_vectors", "NaN")

  data <- reference_data(); data$family_centroids[1, 2] <- Inf
  assert_tox_error(write_flyer(data, path), ERR_NAN_INF, "family_centroids", "Inf")

  data <- reference_data(); data$gene_to_fam <- c(2L, 1L, 4L, 0L, 1L)
  assert_tox_error(write_flyer(data, path), ERR_INVALID_INPUT, "gene_to_fam",
                   "family index above the number of families")

  data <- reference_data(); data$axis_labels <- c("liver", "brain", "liver")
  assert_tox_error(write_flyer(data, path), ERR_INVALID_INPUT, "axis_labels", "repeated axis label")

  data <- reference_data(); data$gene_ids <- c("g1", "g2", "g1", "g4", "g5")
  assert_tox_error(write_flyer(data, path), ERR_INVALID_INPUT, "gene_ids", "repeated gene id")

  data <- reference_data(); data$family_ids <- c("F1", "", "F3")
  assert_tox_error(write_flyer(data, path), ERR_INVALID_INPUT, "family_ids", "empty family id")

  assert_false(file.exists(path), "a refused call creates no file")
}

test_existing_file_is_refused <- function() {
  path <- tempfile(fileext = ".json")
  on.exit(if (file.exists(path)) file.remove(path))
  writeLines("keep me", path)

  assert_tox_error(write_flyer(reference_data(), path), ERR_FILE_OPEN, "filename", "existing file")
  assert_true(identical(readLines(path), "keep me"), "the existing file is untouched")
}

test_shape_mismatch_is_refused_before_the_call <- function() {
  path <- tempfile(fileext = ".json")
  data <- reference_data(); data$gene_ids <- c("g1", "g2", "g3", "g4")
  assert_error(write_flyer(data, path), "gene_ids shorter than the genes")
  assert_false(file.exists(path), "a refused call creates no file")
}

run_all_tests()
